src/recycler.rs

```diff
@@ -0,0 +1,150 @@
+use std::fmt;
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+/// A function that leaves the given type in the same state as Default,
+/// but starts with an existing type instead of allocating a new one.
+pub trait Reset {
+    fn reset(&mut self);
+}
+
+/// An value that's returned to its heap once dropped.
+pub struct Recyclable<T: Default + Reset> {
+    val: Arc<RwLock<T>>,
+    landfill: Arc<Mutex<Vec<Arc<RwLock<T>>>>>,
+}
+
+impl<T: Default + Reset> Recyclable<T> {
+    pub fn read(&self) -> RwLockReadGuard<T> {
+        self.val.read().unwrap()
+    }
+    pub fn write(&self) -> RwLockWriteGuard<T> {
+        self.val.write().unwrap()
+    }
+}
+
+impl<T: Default + Reset> Drop for Recyclable<T> {
+    fn drop(&mut self) {
+        if Arc::strong_count(&self.val) == 1 {
+            // this isn't thread safe, it will allow some concurrent drops to leak and not recycle
+            // if that happens the allocator will end up allocating from the heap
+            self.landfill.lock().unwrap().push(self.val.clone());
+        }
+    }
+}
+
+impl<T: Default + Reset> Clone for Recyclable<T> {
```

Comment:
Seems like using this is an indication of a bug.

Reply:
the window code... it keeps a copy of the blobs while they are referenced in other places.